Feet-related mutations rework

LcsTen opened the pull request:
#### Summary

SUMMARY: Balance "Ankle storage items don't affect feet-related mutations anymore"

#### Purpose of change

Prevent ankle storage items (ankle holster, ankle sheath and ankle ammo pouch) to affect the feet-related mutations (Padded Feet, Roots, Rooter and Tentacle Bracing) by differenciating between shoes/socks/... and items worn at the ankle. Fixes #29540.

#### Describe the solution

The shoes/socks/... and items worn at the ankle are differentiated with the NOT_FOOTWEAR flag.
The function Character::footwear_factor now checks if there are footwear which aren't NOT_FOOTWEAR items.

#### Describe alternatives you've considered

A rather unpopular alternative would be to simply move ankle storage items from feet to legs.

#### Testing

1. Remove the shoes/socks/... you might be wearing
1. Mutate with Padded Feet
1. Check your Base Move Cost (the bonus must apply)
1. Wish for an ankle holster, an ankle sheath and an ankle ammo pouch and wear them all
1. Make sure your Base Move Cost doesn't change (the bonus must still apply)
1. Wear the shoes you worn (or wish for new ones)
1. Make sure your Base Move Cost increased (the bonus must not apply anymore)
1. Remove the shoes you are wearing
1. Mutate with Plant then Rooters
1. Wait for 5 minutes on a Plowable tile
1. Make sure the message "You sink your roots into the soil" appears
1. Wear the shoes your worn and rewait for 5 minutes on a Plowable tile
1. Make sure the message "You sink your roots into the soil" doesn't appear

#### Additional context

As opposed as the issuer told, characters with the Hooves mutation can still wear these items because they suit "abnormally large mutated anatomy".

FuelType-Memes commented:
I think you're looking at it from the wrong angle: it's padded feet that need rework, not ankle storage. Maybe add a flag akin to `ALLOWS_NATURAL_ATTACKS` so it doesn't affect feet mutations?

LcsTen commented:
You're right, I probably overlooked it.

I will try to add a new flag to differentiate the ankle storage items to other items such as socks and shoes, does "DONT_CONNECT_GROUND" sound good?
For now, all the feet-related mutation relies on the Character::footwear_factor, which check if there is any items worn at the feet. I should modify these calls to a new function (maybe Character::ground_factor?), which will check if there is any items worn at the feet which DONT_CONNECT_GROUND.

FuelType-Memes commented:
Maybe `NOT_FOOTWEAR` ?

LcsTen commented:
Sounds good, I'll stick with NOT_FOOTWEAR.
Can I edit this PR, or should I rather make a new one? After all, it shifted from "Move ankle storage items from feet to legs" to "Feet-related mutations rework".

anothersimulacrum commented:
You can edit this one.

Night-Pryanik commented on src/character.cpp:
```diff
@@ -384,6 +384,7 @@ static const std::string flag_SWIMMABLE( "SWIMMABLE" );
 static const std::string flag_SWIM_GOGGLES( "SWIM_GOGGLES" );
 static const std::string flag_UNDERSIZE( "UNDERSIZE" );
 static const std::string flag_USE_UPS( "USE_UPS" );
+static const std::string flag_NOT_FOOTWEAR( "NOT_FOOTWEAR" );
```
Please place it alphabetically.

Night-Pryanik commented on src/character.cpp:
```diff
@@ -9509,7 +9510,7 @@ void Character::update_vitamins( const vitamin_id &vit )
 void Character::rooted_message() const
 {
-    bool wearing_shoes = is_wearing_shoes( side::LEFT ) || is_wearing_shoes( side::RIGHT );
+    bool wearing_shoes = ( ground_factor() == 1.0 );
```
Braces are redundant.

Night-Pryanik commented on src/suffer.cpp:
```diff
@@ -916,7 +916,7 @@ void Character::suffer_from_other_mutations()
     }
 
-    bool wearing_shoes = is_wearing_shoes( side::LEFT ) || is_wearing_shoes( side::RIGHT );
+    bool wearing_shoes = ( ground_factor() == 1.0 );
```
Here too.

LcsTen replied:
Thank you for your feedback!

kevingranade commented:
The sole remaining caller of footwear_factor() is also checking for whether leg/foot mutations are unimpeeded by footwear, so all you really need to do here is make footwear_factor() respect the new flag and make sure it's called consistently.

LcsTen commented:
Good point, I overlooked it for some reasons. I removed the function Character::ground_factor and changed the function Character;:footwear_factor to respect the NOT_FOOTWEAR flag.
Thank you for your feedback.
